spanner/benchmark/ycsb.py

```diff
@@ -0,0 +1,257 @@
+  parser.add_option('-p', '--parameter', action='append', dest='parameters',
+                    default=[], help='The key=value pair of parameter.')
+  parser.add_option('-b', '--num_bucket', action='store', type='int',
+                    dest='num_bucket', default=1000,
+                    help='The number of buckets in output.')
+
+  options, args = parser.parse_args()
+
+  parameters = {}
+  parameters['command'] = args[0]
+  parameters['num_bucket'] = options.num_bucket
+
+  for parameter in options.parameters:
+    parts = parameter.strip().split('=')
+    parameters[parts[0]] = parts[1]
+
+  with open(options.workload, 'r') as f:
+    for line in f.readlines():
+      parts = line.split('=')
+      key = parts[0].strip()
+      if key in OPERATIONS:
+        parameters[key] = parts[1].strip()
+
+  return parameters
+
+
+def open_database(parameters):
+  """Opens a database specified by the parameters from parse_options()."""
+  spanner_client = spanner.Client()
+  instance_id = parameters['cloudspanner.instance']
+  instance = spanner_client.instance(instance_id)
+  database_id = parameters['cloudspanner.database']
+  pool = spanner.BurstyPool(int(parameters['num_worker']))
+  database = instance.database(database_id, pool=pool)
+
+  return database
+
+
+def load_keys(database, parameters):
+  """Loads keys from database."""
+  del KEYS[:]
+  results = database.execute_sql('SELECT u.id FROM %s u' % parameters['table'])
+
+  for row in results:
+    KEYS.append(row[0])
+
+
+def read(database, table, key):
+  """Does a single read operation."""
+  with database.snapshot() as snapshot:
+    result = snapshot.execute_sql('SELECT u.* FROM %s u WHERE u.id="%s"' %
+                                  (table, key))
+    for row in result:
+      key = row[0]
+      for i in range(NUM_FIELD):
+        field = row[i + 1]
+
+
+def update(database, table, key):
+  """Does a single update operation."""
+  field = random.randrange(10)
+  value = ''.join(random.choice(string.printable) for i in range(100))
+  with database.batch() as batch:
+    batch.update(table=table, columns=('id', 'field%d' % field),
+                 values=[(key, value)])
+
+
+def insert(database, table, key):
+  """Does a single insert operation."""
+  raise NotImplementedError('Insert is not implemented.')
+
+
+def scan(database, table, key):
+  """Does a single scan operation."""
+  raise NotImplementedError('Scan is not implemented.')
+
+
+def DoOperation(database, table, operation, latencies_ms):
+  """Does a single operation and records latency."""
+  key = random.choice(KEYS)
+  start = timeit.default_timer()
+  if operation == 'read':
+    read(database, table, key)
+  elif operation == 'update':
+    update(database, table,  key)
+  elif operation == 'insert':
+    insert(database, table, key)
+  elif operation == 'scan':
+    scan(database, table, key)
+  else:
+    raise Exception('Unknown operation: %s' % operation)
```

Comment:
Maybe a more specific exception can be raised?

Reply:
How about ValueError?